File: rag_pipeline.js
const SimpleVectorStore = require('./vector2');
const axios = require('axios');

class RAGPipeline {
    constructor() {
        this.vectorStore = new SimpleVectorStore();
        this.initialized = false;
        
        // Ollama configuration (matching your Python setup)
        this.aiModelUrl = 'http://127.0.0.1:11434/api/generate';
        this.modelName = 'llama3.2:1b'; // Match your Python model
    }

    async initialize() {
        if (!this.initialized) {
            await this.vectorStore.initialize();
            this.initialized = true;
        }
    }

    async callLocalAI(prompt) {
        try {
            console.log('🤖 Calling local AI model...');
            
            const response = await axios.post(this.aiModelUrl, {
                model: this.modelName,
                prompt: prompt,
                stream: false,
                options: {
                    temperature: 0.7,
                    num_predict: 500
                }
            }, {
                timeout: 60000,
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            if (response.data && response.data.response) {
                return response.data.response.trim();
            } else {
                throw new Error('Invalid response format from AI model');
            }

        } catch (error) {
            console.error('❌ Error calling local AI model:', error.message);
            throw new Error(`Unable to connect to local AI model: ${error.message}`);
        }
    }

    createPrompt(userQuestion, contextQA) {
        // Create context from Q&A pairs (matching Python logic exactly)
        let context = "";
        
        contextQA.forEach((result, index) => {
            const metadata = result.metadata;
            const question = metadata.question;
            const answer = metadata.answer;
            const source = metadata.source;
            // Format exactly like Python version
            context += `Source: ${source}\nQ: ${question}\n\nA: ${answer}\n\n`;
        });

        // Use the EXACT same prompt template as your Python version
        const prompt = `You are a helpful AI assistant for the platform team's knowledge base. Answer questions based ONLY on the provided context.

Be concise and helpful. If you don't have enough information, say so clearly.

Context from knowledge base:
${context}

Question: ${userQuestion}

Answer:`;

        return prompt;
    }

    async getBotResponse(query) {
        try {
            await this.initialize();
            
            console.log(`🔍 Searching for relevant questions for: "${query}"`);
            
            // Use the question search method (k=5, matching Python)
            const relevantQuestions = await this.vectorStore.questionSearch(query, 5, 0.3);
            
            if (relevantQuestions.length === 0) {
                return "Sorry, I couldn't find relevant information in the knowledge base for your question. Please contact the platform team directly.";
            }

            console.log(`📚 Found ${relevantQuestions.length} relevant questions`);
            
            // Create prompt with context
            const prompt = this.createPrompt(query, relevantQuestions);
            
            // ALWAYS try to call the AI model first
            console.log('🤖 Generating response with local AI model...');
            
            try {
                const aiResponse = await this.callLocalAI(prompt);
                
                // Get unique sources (matching Python logic)
                const sources = relevantQuestions.slice(0, 3).map(result => result.metadata.source);
                const uniqueSources = [...new Set(sources)];
                
                let sourcesInfo = "";
                if (uniqueSources.length > 0) {
                    sourcesInfo = `\n\n📚 Sources: ${uniqueSources.slice(0, 2).join(', ')}`;
                }
                
                const fullResponse = aiResponse + sourcesInfo;
                
                console.log('✅ AI response generated successfully');
                return fullResponse;
                
            } catch (aiError) {
                console.error('❌ AI model error:', aiError.message);
                console.log('📋 Falling back to direct context response');
                
                // Fallback: provide a structured answer from the context
                const bestMatch = relevantQuestions[0];
                const metadata = bestMatch.metadata;
                
                let fallbackResponse = `Based on the knowledge base:\n\n${metadata.answer}`;
                
                if (metadata.source) {
                    fallbackResponse += `\n\n📚 Source: ${metadata.source}`;
                }
                
                return fallbackResponse;
            }

        } catch (error) {
            console.error('❌ Error in RAG pipeline:', error);
            return `Sorry, I encountered an error: ${error.message}`;
        }
    }

    // Check if the AI model is available
    async checkModelStatus() {
        try {
            const response = await axios.get('http://localhost:11434/api/tags', {
                timeout: 5000
            });
            
            const models = response.data.models || [];
            const hasModel = models.some(model => model.name.includes(this.modelName));
            
            if (hasModel) {
                console.log(`✅ AI model ${this.modelName} is available`);
                return true;
            } else {
                console.log(`❌ AI model ${this.modelName} not found. Available models:`, models.map(m => m.name));
                return false;
            }
            
        } catch (error) {
            console.log(`❌ Cannot connect to Ollama server: ${error.message}`);
            return false;
        }
    }
}

// Create singleton instance
const ragPipeline = new RAGPipeline();

async function getBotResponse(query) {
    try {
        // Check model status first
        const modelAvailable = await ragPipeline.checkModelStatus();
        if (!modelAvailable) {
            console.log('⚠️  AI model not available, will use fallback in getBotResponse');
        }
        
        return await ragPipeline.getBotResponse(query);
    } catch (error) {
        console.error('❌ Error in getBotResponse:', error);
        return `Sorry, I encountered an error: ${error.message}`;
    }
}

module.exports = { getBotResponse };

File: rag_pipeline.py
# rag_pipeline.py

from langchain_ollama.llms import OllamaLLM
from langchain_core.prompts import ChatPromptTemplate
from vector2 import retrieval  # Your existing retrieval logic


# Initialize model
model = OllamaLLM(model="llama3.2:1b")

# Create prompt template
template = """
You are a helpful AI assistant for the platform team's knowledge base. Answer questions based ONLY on the provided context.

Be concise and helpful. If you don't have enough information, say so clearly.

Context: {context}
Question: {question}

Answer:"""

template_with_context = """
You are a helpful AI assistant for the platform team's knowledge base. Answer questions based ONLY on the provided context.

IMPORTANT: This conversation is part of a thread. Below is the conversation history that you MUST consider when answering:

=== CONVERSATION HISTORY ===
{thread_context}
=== END CONVERSATION HISTORY ===

Use this conversation history to:
1. Understand the context and continuation of the discussion
2. Refer back to previous questions and answers when relevant
3. Provide follow-up answers that build on the conversation
4. Clarify or expand on previous responses if needed

Be concise and helpful. If you don't have enough information, say so clearly.

Knowledge Base Context: {context}
Current Question: {question}
put in your mind that I can not see the knowledge base i can only see this question {question} and this conversation history so i need a detailed sumarized answer i do not want you to list the context


Answer:"""

template_no_context = """
You are a helpful AI assistant for the platform team's knowledge base. Answer questions based ONLY on the provided context.

Be concise and helpful. If you don't have enough information, say so clearly.

Context: {context}
Question: {question}
put in your mind that I can not see the knowledge base i can only see this question {question} so i need a detailed sumarized answer i do not want you to list the context


Answer:"""

prompt = ChatPromptTemplate.from_template(template)
chain = prompt | model

prompt_with_context = ChatPromptTemplate.from_template(template_with_context)
prompt_no_context = ChatPromptTemplate.from_template(template_no_context)

chain_with_context = prompt_with_context | model
chain_no_context = prompt_no_context | model

def get_bot_response_with_context(message, thread_context=""):
    """Get response from RAG pipeline with optional thread context"""
    try:
        print(f"Processing message: {message}")
        if thread_context:
            print(f"Thread context length: {len(thread_context)} characters")
        
        # Get documents from your existing retrieval
        docs = retrieval.invoke(message)
        
        if not docs:
            if thread_context:
                # Even without knowledge base docs, we can respond using thread context
                return f"I don't have specific information in the knowledge base for this question, but based on our conversation, I can see we were discussing related topics. Please contact the platform team for more detailed information."
            else:
                return "Sorry, I couldn't find relevant information in the knowledge base for your question. Please contact the platform team directly."
        
        # Prepare context (limit to prevent too long responses)
        context_parts = []
        for doc in docs[:5]:  # Limit to 5 most relevant docs
            source = doc.metadata.get('source', 'Unknown')
            content = doc.page_content[:400]  # Limit content length for Slack
            context_parts.append(f"Source: {source}\n{content}")
        
        context = "\n\n".join(context_parts)
        
        # Generate response using appropriate chain
        if thread_context.strip():
            # Use context-aware prompt
            result = chain_with_context.invoke({
                "thread_context": thread_context,
                "context": context, 
                "question": message
            })
        else:
            # Use regular prompt
            result = chain_no_context.invoke({
                "context": context, 
                "question": message
            })
        
        # Add source information
        sources = [doc.metadata.get('source', 'Unknown') for doc in docs[:3]]
        unique_sources = list(set(sources))
        
        if len(unique_sources) > 0:
            sources_text = f"\n\n📚 Sources: {', '.join(unique_sources[:2])}"
            result += sources_text
        
        return result
        
    except Exception as e:
        print(f"Error in get_bot_response_with_context: {str(e)}")
        return f"Sorry, I encountered an error: {str(e)}"

def get_bot_response(message):
    """Get response from RAG pipeline"""
    try:
        print(f"Processing message: {message}")
        
        # Get documents from your existing retrieval
        docs = retrieval.invoke(message)
        
        if not docs:
            return "Sorry, I couldn't find relevant information in the knowledge base for your question. Please contact the platform team directly."
        
        # Prepare context (limit to prevent too long responses)
        context_parts = []
        for doc in docs[:5]:  # Limit to 5 most relevant docs
            source = doc.metadata.get('source', 'Unknown')
            content = doc.page_content[:400]  # Limit content length for Slack
            context_parts.append(f"Source: {source}\n{content}")
        
        context = "\n\n".join(context_parts)
        
        # Generate response using your model
        result = chain.invoke({"context": context, "question": message})
        
        # Add source information
        sources = [doc.metadata.get('source', 'Unknown') for doc in docs[:3]]
        unique_sources = list(set(sources))
        
        if len(unique_sources) > 0:
            sources_text = f"\n\n📚 Sources: {', '.join(unique_sources[:2])}"
            result += sources_text
        
        return result
        
    except Exception as e:
        print(f"Error in get_bot_response: {str(e)}")
        return f"Sorry, I encountered an error: {str(e)}"